### ISSUE TITLE
Question about neighbour computation

### ISSUE THREAD

**Issue by ayushjain1144** [HuguesTHOMAS/KPConv-PyTorch#104]
Hi,

Thanks for your awesome work! 

I wanted to understand how you compute neighbouring points for the kernel, especially in the deformable case? My understanding is that the rigid KPConv would produce some offsets for each kernel points during the forward pass, and then you would need to find points in a radius r around the kernel+offset points. Since offset points are produced by a neural network during the forward pass, my understanding is that its not possible to use precomputed neighbours, but from the code, it looks like you are somehow able to use precomputed neighbours. Could you help me understand it better? I think that computing neighbours during the forward pass would be very expensive, so if we can somehow use the precomputed neighbours, that would be pretty important. 

Thanks!
Ayush

**Comment by HuguesTHOMAS** [comment]
We just define a bigger radius for deformable neighbors, that allows the deformed points to go further than the original radius. It is fixed in advance which is why we can precompute the neighbors:

The deformable radius is defined by the parameter `config.deform_radius`, while the original radius is defined by `config.conv_radius`
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/1cd3742fc57e5ba9d21f7909c78a05a971781351/train_S3DIS.py#L105-L109

**Comment by ayushjain1144** [comment]
Hi @HuguesTHOMAS, thanks for your reply!

I am still not very clear with this. My understanding is that you need to find neighbouring points belonging to the original pointcloud from the "predicted offset point". So I am imagining that we need to find points belonging to original scan that lie in a radius R with the offset point at center. Since this offset point is only available after forward pass, how does fixing a bigger radius in advance help? 

Thank you so much!

**Comment by HuguesTHOMAS** [comment]
Here is what happens:
1) compute the neighbouring points belonging to the original pointcloud from the center of the kernel (with a larger radius)
2) Compute the distances between all these neighbors to all the offset kernel point.
3) Get influence value from this distance (influence value is zero if distance is superior to the influence distance `KP_extent`)

As you can see, fixing a bigger radius is crucial here, because if an offset kernel point is located outside this radius, it will not get its neighbors correctly.

Now for your follow-up question, the regularization loss applied to the offset kernel points forces them to get to the closest original point cloud point. But this means with our implementation, the closest point from the neighbors computed with the bigger radius. As a result, the kernel points are soflty constrained to remain in this big radius

Best,
Hugues

**Comment by ayushjain1144** [comment]
Hi Hugues, 

Thanks for the great explanation. Now I fully understand what's happening. In my native implementation, I was trying to compute neighbours after each forward pass. That made it slow as well as memory intensive. But one thing I noticed was that the model worked equally well with and without the regularisation losses.  I think that with your implementation, the regularisation losses would play a much more important part. 

Thanks a lot for all your help!!
